[Week1] ToDo REST API 만들기

JIY0UNG opened the pull request:
일단 handler가 없는 상태의 서버만 만들어봤습니다!
pull request가 익숙하지 않아서 이게 맞는 건지 잘 모르겠지만 일단 한번 해봤어요😂

JIY0UNG committed:
Handler가 없는 상태의 서버 생성

johngrib reviewed (commented):
만나서 반갑습니다. PR에 익숙하지 않다고 하셨지만 잘 올리셨네요! 아샬님 강의를 수강하시면서 조금씩 완성해서 PR을 계속 갱신해 보세요.

JIY0UNG committed:
사용하지 않는 getGreeting 메서드 삭제

JIY0UNG committed:
요청을 처리할 TodoHttpHandler 추가

JIY0UNG committed:
TodoHttpHandler에 task 목록을 얻어오는 코드 추가

GET 요청을 받았을 경우, 이에 대한 응답으로 json 형태의 task 목록을 보내주도록 함

JIY0UNG committed:
TodoHttpHandler에 task를 생성하는 코드 추가

JIY0UNG commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -31,9 +32,19 @@ public void handle(HttpExchange exchange) throws IOException {
+                    tasks.add(task); // 리스트에 task 추가
+                }
+            }
         }
```
과제에서 처리해야 하는 요청들의 path가 모두 /tasks라서 상위 if문으로 묶어서 처리하였는데, 작성하고 보니 if문을 3번 중첩한 게 보기 안 좋은 것 같네요.. 혹시 좋은 방법 없을까요??

hannut91 replied:
음 아마도 중복을 제거하기 위해서 시도를 하신 것 같아요. if문이 중복되긴 하지만, 여기서는 어쩔 수 없는 것 같아요. 대부분의 웹 프레임 워크들이 경로와 메서드로 구분을 해서, 앞으로 이렇게 할 경우는 없을 것 같아요.

JIY0UNG replied:
넵 과제를 좀 더 진행하면서 개선해 보도록 하겠습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,23 @@
 
     public static void main(String[] args) {
+        try {
+            InetSocketAddress address = new InetSocketAddress(8000);
+            HttpServer httpServer = HttpServer.create(address, 0);
```
8000은 포트번호를 의미하고 있는 것 같아요. 8000이 단순히 숫자가 아니라 특별한 의미를 갖는다는 것을 상수로 표현해보는 것은 어떨까요?

See also
  - [Replace Magic Literal](https://refactoring.com/catalog/replaceMagicLiteral.html)

JIY0UNG replied:
앗 그러네요 해당 숫자의 의미를 명확히 알 수 있도록 하는 게 좋겠군요! 수정하도록 하겠습니다~~

hannut91 reviewed (commented):
> 일단 한번 해봤어요

라는 말이 인상적이네요! 앞으로도 많은 시도를 해보셨으면 좋겠습니다!

JIY0UNG committed:
기존에 포트번호를 직접 썼던 것을 상수로 변경

변하지 않는 값인 포트번호에 대해 그 의미를 명확히 알 수 있도록 상수로 선언함

JIY0UNG committed:
POST 요청 시 생성된 task 정보를 클라이언트 측에 넘겨주는 코드 추가

JIY0UNG committed:
특정 할 일의 상세 정보를 조회하는 코드 추가

GET 요청 uri에 task id가 포함되어있는 경우, tasks 리스트에서 task id에 해당하는 task를 가져오도록 함

JIY0UNG committed:
할 일의 제목을 수정하는 코드 추가

PUT 요청 시 uri에 포함되어 있는 task id로 tasks 리스트에서 해당 task를 가져온 후, 새로 입력한 제목으로 setTitle을 수행하도록 함

JIY0UNG committed:
할 일을 삭제하는 코드 추가

DELETE 요청 시 uri에 포함되어 있는 task id로 tasks 리스트에서 해당 task를 remove 하도록 함

JIY0UNG committed:
tasks 리스트에서 특정 task를 가져오는 코드 수정

remove 함수를 사용하면 삭제 후 한 칸씩 앞으로 당겨지므로 get(taskId - 1)으로 가져온 객체가 원했던 객체와 다른 경우가 생긴다는 것을 발견함. 그래서 반복문으로 tasks를 탐색해서 id가 일치하는 객체를 가져오는 방식으로 수정하였음.

JIY0UNG commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
-    }
 
     public static void main(String[] args) {
+        final int PORT_NUM = 8000;
+
+        try {
+            InetSocketAddress address = new InetSocketAddress(PORT_NUM);
```
포트번호를 상수로 표현해보라는 윤석님의 리뷰를 보고 수정하였습니다!

JIY0UNG replied:
포트번호를 상수로 표현해보라는 윤석님의 리뷰를 보고 수정하였습니다! 8000이라는 숫자가 뭘 의미하는지 명확하게 표현되어서 좋은 것 같아요

JIY0UNG replied:
포트번호를 상수로 표현해 보라는 윤석님의 리뷰를 보고 수정하였습니다! 8000이라는 숫자가 뭘 의미하는지 명확하게 표현되어서 좋은 것 같아요

JIY0UNG commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,116 @@
+                if(!body.isBlank()) { // request body가 비어있지 않으면
+                    Task task = toTask(body); // body의 내용을 Task 객체로 만듦
+                    tasks.add(task); // 리스트에 task 추가
+                    content = task.toString(); // 내가 생성한 task 정보를 출력
```
클라이언트 측에 생성한 task 정보를 출력할 수 있도록 코드를 추가하였습니다

JIY0UNG commented:
안녕하세요~~ 오늘은 ToDo 상세 조회, 제목 수정, 삭제 기능을 구현해 보았습니다
중복되는 코드가 많아서 좀 지저분해 보이네요😂
오늘은 너무 피곤해서 리팩토링은 내일로 미루고.. 작업한 것까지만 먼저 pr 올려봅니다!
리뷰 부탁드려요😊

JIY0UNG reviewed (commented):
안녕하세요~~ 오늘은 ToDo 상세 조회, 제목 수정, 삭제 기능을 구현해 보았습니다
중복되는 코드가 많아서 좀 지저분해 보이네요😂
오늘은 너무 피곤해서 리팩토링은 내일로 미루고.. 작업한 것까지만 먼저 pr 올려봅니다!
리뷰 부탁드려요😊

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
-    }
 
     public static void main(String[] args) {
+        final int PORT_NUM = 8000;
```
fianl static 인 것에만 UPPER_SNAKE_CASE로 작성하는 관례가 있습니다. 그리고 `num`보다는 `NUMBER`라고 써주는게 더 좋을 것 같습니다

JIY0UNG replied:
> fianl static 인 것에만 UPPER_SNAKE_CASE로 작성하는 관례가 있습니다.

앗 제가 잘못 알고 있었던 것 같네요. 알려주셔서 감사합니다!

JIY0UNG replied:
그리고 말씀하신 final static을 보니 갑자기 static과 final static의 차이에 대해 헷갈리기 시작해서 자료를 좀 찾아보았습니다.

final로 선언과 함께 초기화 시키면 상수로 만들 수도 있긴 하지만, 본질적으로 final은 객체별로 다른 값을 할당할 수도 있는 키워드이기 때문에, 상수를 선언하는 데에는 모든 객체가 동일한 값을 갖는 static final을 사용하는 게 좋겠다는 결론을 내리게 되었습니다.

따라서 기존에 final로 선언했던 포트 번호를 static final로 바꾸려고 하는데, 사실 별 상관없는 건지, 바꾸는 게 더 나은 건지 확신이 안 서서 의견 좀 여쭙고 싶습니다

hannut91 replied:
지금 상황에서는 사실 중요하지는 않아요 ㅎㅎ 이 객체를 다르게 만들 경우도 없을 것 같고, 지금 단계에서도 충분히 좋은 코드입니다.

JIY0UNG replied:
그렇군요 좀 더 큰 프로젝트를 할 때 적용해봐야겠네요! 답변 감사드립니다~~

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,116 @@
+
+        if(path.startsWith("/tasks")) { // path가 /tasks로 시작하는 요청들에 대한 처리
+            // 할 일 목록 얻기
+            if(method.equals("GET")) {
+                content = tasksToJSON(); // 리스트에 저장되어 있던 task들을 json 형태로 가져오기
+            }
+
+            // 상세 조회하기
+            if(method.equals("GET")) {
+                if(path.length() > 6) {
+                    int taskId = Integer.parseInt(path.substring(path.lastIndexOf('/') + 1)); // path에 포함되어있는 task id 추출
+
+                    for(Task task : tasks) {
+                        if(task.getId() == taskId) {
+                            content = task.toString();
+                            break;
+                        }
+                    }
+                }
+            }
```
`method.equals("GET")`가 중복되어 실행되고 있네요. 정리가 필요해 보입니다

JIY0UNG replied:
넵 중복 코드가 없도록 별도 함수로 분리하는 작업을 해야겠어요

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,116 @@
+                            break;
+                        }
+                    }
+                }
+            }
+
+            // 할 일 생성하기
+            if(method.equals("POST")) {
+                if(!body.isBlank()) { // request body가 비어있지 않으면
+                    Task task = toTask(body); // body의 내용을 Task 객체로 만듦
+                    tasks.add(task); // 리스트에 task 추가
+                    content = task.toString(); // 내가 생성한 task 정보를 출력
+                }
+            }
+
+            // 할 일 제목 수정하기
+            if(method.equals("PUT")) {
+                int taskId = Integer.parseInt(path.substring(path.lastIndexOf('/') + 1)); // path에 포함되어있는 task id 추출
+                Task updateTask = toTask(body);
+
+                for(Task task : tasks) {
+                    if(task.getId() == taskId) {
+                        task.setTitle(updateTask.getTitle());
+                        content = task.toString();
+                        break;
+                    }
+                }
+            }
+
+            // 할 일 삭제하기
+            if(method.equals("DELETE")) {
+                int taskId = Integer.parseInt(path.substring(path.lastIndexOf('/') + 1)); // path에 포함되어있는 task id 추출
+                for(Task task : tasks) {
+                    if(task.getId() == taskId) {
+                        tasks.remove(task);
+                        break;
+                    }
+                }
+            }
+        }
```
조건문이 길어지다보니, 앞의 조건에 부합한다고 하더라도, 함수의 진행이 끝나지 않고 계속 되다보니, 사실은 필요없는 코드들도 계속 읽어야 할 수 밖에 없었어요.

만약 ~한 조건일 때 필요한 일만하고 빠르게 함수를 종료한다면, 내가 필요한 코드만 빠르게 읽을 수 있을 것 같아요.

JIY0UNG replied:
저도 if문이 너무 길다고 생각했어서, 아래 코드와 같이 함수로 분리하는 작업을 하고 있습니다!
```java
case "GET":
    if(path.length() > 6) {
        taskId = getTaskId(path);
        content = getOneTask(taskId);
    }
    else {
        content = getAllTask();
    }
    break;
```

JIY0UNG replied:
> 사실은 필요없는 코드들

이 말이 잘 이해가 되지 않는데, if문 내에서 여러 가지 일을 모두 처리하고 있었기 때문에 굳이 보지 않아도 되는 세부 처리 과정들까지 보게 되었다는 말씀이실까요??

hannut91 replied:
예를들어서 요청한 메서드가 GET메서드라면 할 일 목록을 응답하고 있죠. 내가 GET메서드에 대한 처리만 읽고 있었다면, 이 부분만 읽고 싶었을 겁니다. 그래서 GET메서드 처리 후에 함수가 바로 반환을 했더라면, 그 아래의 코드들은 읽을 필요가 없으니 읽는 사람에게 부하가 줄어들게 돼요

See also
  - https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,116 @@
+    // request body의 내용을 Task 객체로 만들기
+    private Task toTask(String content) throws JsonProcessingException {
+        Task task = objectMapper.readValue(content, Task.class);
+        task.setId(id++); // 할 일을 새로 생성할 때마다 id가 증가하도록 함
```
후위연산자를 사용해서 값이 변경되기 전의 값을 할당하고 나서 값을 증가시키고 있는 것 같아요. 사용법을 정확히 알고 있다면 `++value` 와 `value++` 을 구분해서 사용할 수 있겠지만, 사실 조금 헷갈릴수도 있는 문법이에요. 우리가 한 번 더 생각해야되는거죠. 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

JIY0UNG replied:
별생각 없이 후위 연산자를 사용했었는데, 리뷰를 보고 다시 생각해 보게 되었습니다.
위 코드에서는 전위 연산자를 사용하여 이미 증가된 값을 할당하는 게 더 자연스러운 처리 순서가 되겠군요!

JIY0UNG replied:
> 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

증감연산자를 사용하는 게 뭔가 더 깔끔해 보여서 그동안 증감연산자를 선호해왔는데, 이런 상황에서는 헷갈릴 수 있으니 아래 코드와 같이 작성하는 게 더 좋다고 이해하면 될까요??
```java
id = id + 1;
task.setId(id);
```

hannut91 replied:
네 맞습니다 ㅎㅎ

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,28 @@
+        this.title = title;
+    }
+
+    @Override
+    public String toString() {
+        return "{\"id\":" + id + ",\"title\":" + title + "\"}";
+    }
```
String.format을 사용해서 표현하는 건 어떨까요? 다음과 같이 표현할 수 있을거에요

```java
return String.format("{ id = %s, title = %s }", id, title);
```

더 좋은건 Template literal이지만 자바에서는 지원하질 않네요 ㅠ 만약 이런게 지원하는 코틀린을 사용했다면 다음과 같이 사용할 수 있을거에요

```java
return "{ id = $id, title = $title }"
```

JIY0UNG replied:
오 C언어에서만 이런 식으로 출력 가능한 줄 알았는데, 아니었군요
제가 썼던 코드보다 String.format을 사용한 코드가 훨씬 가독성이 좋네요!!

hannut91 commented:
오늘도 화이팅 입니다 ㅎㅎ

JIY0UNG committed:
toString 메서드 반환문의 가독성을 높이기 위해 String.format 사용

JIY0UNG committed:
각 HTTP 메서드에 대한 처리를 별도 함수로 분리

JIY0UNG reviewed (commented):
안녕하세요~!
오늘은 각 HTTP 메소드에 대한 처리를 별도 함수로 분리하는 작업을 했습니다. 리뷰 부탁드립니다!!

상수나 증감연산자에 대해서는 윤석님 의견을 좀 더 들어볼 수 있을까 해서 아직 커밋을 하지 않았어요. 혹시 시간 나실 때 위에 제가 남긴 댓글들도 한번 봐주시면 정말 감사하겠습니다!
오늘도 좋은 하루 되세요😄😄

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,153 @@
+        int taskId;
+
+        if(path.startsWith("/tasks")) { // path가 /tasks로 시작하는 요청들에 대한 처리
+            switch (method) {
```
`switch`문을 쓸 때는 반드시 `default`를 사용하는게 좋습니다.

JIY0UNG replied:
예외 상황을 가정하지 않고 코드를 작성했었는데, GET, POST, PUT, PATCH, DELETE 외의 다른 메서드를 사용한 요청이 들어왔을 경우를 default에서 처리해주는 것이 더 안전하겠군요!

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,153 @@
+                case "PUT":
+                case "PATCH":
+                    taskId = getTaskId(path);
+                    Task updateTask = toTask(body);
```
`updateTask`는 무언가를 수정해! 라고 하는 것 같아요. 실제로 같은 이름의 메서드가 아래에 있어서 더 헷갈리는 것 같습니다

JIY0UNG replied:
윤석님 리뷰를 보고 나니까 updateTask는 변수명보다는 메서드명으로 더 적절한 것 같네요.
게다가 메서드명과 변수명을 동일하게 짓는 이런 실수를 하다니..ㅎㅎ 좀 더 좋은 이름을 고민해 보겠습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,153 @@
+
+    // 할 일 제목 수정하기
+    private String updateTask(int taskId, String updateTitle) {
+        String returnTask = null;
```
반환할 할 일이라서 `returnTask`라고 하신 것 같아요. 실제로 사용하는 곳을보면 `title`로 사용하고 있던데, 이 값이 정확히 무엇을 의미하는지 드러내면 좋을 것 같아요

JIY0UNG replied:
넵 해당 변수에는 id와 title을 나타내는 문자열이 저장되니까 이름을 idAndTitle로 수정하겠습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    // 할 일 제목 수정하기
+    private String updateTask(int taskId, String updateTitle) {
```
`tasks`에서 우리가 원하는 task를 찾지 못할 경우 `null`을 반환하고 있는 것 같아요.이 함수를 사용하는 입장에서는 이 함수가 `null`을 반환할 수 있는지 예상할 수 없고, 그리고 `null`처리를 하지 않아도 자바 컴파일러는 아무런 불만도 표현하지 않기 때문에 위험성이 있는 것 같아요. 만약 반환 타입을 Optional 그대로 반환한다면, 함수의 Signature에서 이 값이 있을 수도 있고 없을 수도 있다는 의도를 드러내고 있고, 없는 경우를 처리하지 않으면 컴파일도 되지 않아서 에러 처리를 강제할 수 있어요

하지만 Optional 반환은 다른 부분이 어떻게 만들어졌냐가 관건이라고 보는데, 함수형으로 처리 중이 아니라면 예외가 가장 일반적이죠. Optional을 리턴하면 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에 사용하기 힘든 인터페이스가 될 가능성이 커요.

JIY0UNG replied:
task를 찾지 못한 경우 null을 반환하고 있어서 좋지 못하다는 건 이해했는데, 그 이후에 하신 말씀이 잘 이해가 안 됩니다ㅠ

task를 1개만 POST 한 후에 task id가 2인 task의 title을 변경하려고 할 때(원하는 task를 찾지 못한 경우)
터미널 창에 아래와 같은 에러 메시지가 뜨는데,
혹시 이게 지금 null을 반환하고 있어서 이런 오류가 발생하는 걸까요??
![image](https://user-images.githubusercontent.com/98736689/167280023-eb212790-4c75-4e04-a2eb-2921d7d7faab.png)

JIY0UNG replied:
> 만약 반환 타입을 Optional 그대로 반환한다면, 함수의 Signature에서 이 값이 있을 수도 있고 없을 수도 있다는 의도를 드러내고 있고

이 말이 잘 이해가 되지 않습니다.. 음 지금 제가 Optional 반환을 하고 있는 건가요??

JIY0UNG replied:
> 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에

이 말도 무슨 뜻인지 잘 모르겠네요ㅠㅠ

JIY0UNG committed:
포트번호 변수명 변경

- UPPER_SNAKE_CASE는 final static에만 사용
- num보다 좀 더 명확한 표현인 number로 수정

JIY0UNG committed:
헷갈릴 수 있는 증감 연산자를 사용하는 대신 연산을 풀어서 작성

JIY0UNG committed:
path가 /tasks로 시작하는 요청들을 처리하는 함수를 별도로 만듦

- 가독성을 높이기 위해 각 조건문 만족 시 그에 따른 일을 수행한 후, 다른 구문이 수행되기 전에 함수를 return 하도록 함

JIY0UNG committed:
변수가 어디에 쓰이는 지 좀 더 명확히 드러낼 수 있는 이름으로 수정

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,153 @@
+        if(path.startsWith("/tasks")) { // path가 /tasks로 시작하는 요청들에 대한 처리
+            switch (method) {
+                case "GET":
+                    if(path.length() > 6) {
```
왜 6인지 이해하기 어렵네요. 아마도 의도를 더 명확하게 드러내려면 어떻게 해야 할까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    // 상세 조회
+    private String getOneTask(int taskId) {
+        String returnTask = null;
+
+        for(Task task : tasks) {
+            if(task.getId() == taskId) {
+                returnTask = task.toString();
+                break;
+            }
+        }
+
+        return returnTask;
+    }
```
스트림을 이용해서 문제를 해결해 보는 것은 어떨까요?
